diff --color -urN alloc_backup/src/lib.rs alloc/src/lib.rs
--- alloc_backup/src/lib.rs	2025-03-18 22:41:47
+++ alloc/src/lib.rs	2025-03-17 18:14:44
@@ -245,7 +245,10 @@
 pub mod slice;
 pub mod str;
 pub mod string;
-#[cfg(all(not(no_rc), not(no_sync), target_has_atomic = "ptr"))]
+#[cfg(any(
+    all(not(no_rc), not(no_sync), target_has_atomic = "ptr"),
+    all(target_arch = "arm", target_os = "none", target_abi = "eabi", not(target_has_atomic = "ptr")),
+))]
 pub mod sync;
 #[cfg(all(not(no_global_oom_handling), not(no_rc), not(no_sync)))]
 pub mod task;
